elastic/eui
# Upgrade to Node 16

### Summary

This PR attempts to upgrade our Node version from 14.x to 16.x to match Kibana.

The devDependencies that were preventing us from quickly and easily doing so were:

- `nodegit`
  - see: https://github.com/nodegit/nodegit/issues/1840
  - nodegit currently has a `0.28` alpha that works with v16 that I've opted to point our repo at. I feel relatively confident doing so since their alpha has been out for a while now, and nodegit only affects dev processes / our release script and not production code.
  - This should not affect CI since nodegit is not called on CI
- `node-sass`
  - node-sass has been updated to work with v16 since their 6.x release and above.
  - see: https://github.com/sass/node-sass/issues/3077
  - A variety of our Sass dependencies require/point to older versions of `node-sass`:
    - `sass-loader`: Updated to work with node-sass v6/Node v16 in [10.2.0](https://github.com/webpack-contrib/sass-loader/releases/tag/v10.2.0)
      - Note that their _latest_ release is technically at 12.x (which does support node v16), but unfortunately we cannot upgrade to 11.x+ because that would require us to use Webpack 5
    - `sass-extract`: see https://github.com/jgranstrom/sass-extract/issues/47
    - `sass-vars-to-js-loader`
    - `sass-lint` (TBH, we should probably be using `stylelint` instead)
    - `sass-lint-auto-fix` (see above, also not sure when the last time was that we used this)
    - To get all of the above to use the correct/latest version of node-sass, I've made use of [Yarn's `resolutions` field](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/).

## QA

Setup:
- Checkout this branch
- `nvm use` (don't forget this step!)
- `npm rebuild` OR `rm -rf node_modules` (without this step, I believe you'll get this error):
```
Error: The module 'eui/node_modules/nodegit/build/Release/nodegit.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 83. This version of Node.js requires
NODE_MODULE_VERSION 93. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```
- `yarn`

Ensure the below scripts all work as expected/without errors:
- `node-sass` related scripts
  - [x] yarn start
  - [x] yarn lint-sass
  - [x] yarn lint-sass-fix
  - [x] yarn compile-scss
    - [x] Confirm that the output `.css` files on this branch are the same as the output CSS on main using diffchecker.com
  - Extra scripts testing just in case
    - [x] yarn test-cypress
    - [x] yarn build
- `nodegit` related scripts
  - [x] yarn update-token-changelog
  - [x] npm run release --dry-run
    - NOTE: To test on this branch, you'll have to open `scripts/release.js`, comment out `await ensureMainBranch();` on line 42, and temporarily commit that to have a clean working directory

### Checklist

N/A, this is a dev only change - the above QA steps should suffice.
